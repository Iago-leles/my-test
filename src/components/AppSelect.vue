<template>
  <div>
    <span class="text-subtitle1 text-grey-6 text-weight-bold">
      {{ label }}

      <span v-if="required" class="text-red-5">*</span>
    </span>

    <q-select
      :model-value="modelValue"
      use-input
      hide-selected
      fill-input
      dense
      outlined
      clearable
      input-debounce="0"
      :options="filteredOptions"
      @filter="filterFn"
      @update:model-value="$emit('update:modelValue', $event)"
      style="width: 100%; padding-bottom: 32px"
    >
      <template v-slot:no-option>
        <q-item>
          <q-item-section class="text-grey"> Nenhum resultado encontrado </q-item-section>
        </q-item>
      </template>
    </q-select>
  </div>
</template>

<script setup lang="ts">
import { ref, computed } from 'vue'

interface Option {
  label: string
  value: string | number
}

const props = defineProps<{
  modelValue: Option | null
  label: string
  required?: boolean
  options: Array<Option>
}>()

const originalOptions = ref<Array<Option>>(props.options)

const formattedOptions = computed(() => {
  return originalOptions.value.map((option) => {
    if (typeof option === 'string') {
      return { label: option, value: option }
    }
    return {
      label: option.label || (typeof option === 'object' ? JSON.stringify(option) : String(option)),
      value: option.value,
    }
  })
})

const filteredOptions = ref<Array<Option>>(formattedOptions.value)

const filterFn = (val: string, update: (fn: () => void) => void) => {
  update(() => {
    if (val === '') {
      filteredOptions.value = formattedOptions.value
      return
    }

    const needle = val.toLowerCase()
    filteredOptions.value = formattedOptions.value.filter((v) => {
      const label = v.label?.toLowerCase() || ''
      return label.indexOf(needle) > -1
    })
  })
}
</script>
