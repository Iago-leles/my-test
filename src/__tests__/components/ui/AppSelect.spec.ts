import { mount, flushPromises } from '@vue/test-utils'
import { expect, describe, it } from 'vitest'
import AppSelect from '@/components/ui/AppSelect.vue'

describe('AppSelect', () => {
  const options = [
    { label: 'Opção 1', value: 1 },
    { label: 'Opção 2', value: 2 },
    { label: 'Opção 3', value: 3 },
  ]

  it('should match snapshot', () => {
    const wrapper = mount(AppSelect, {
      props: {
        modelValue: null,
        label: 'Test Select',
        options,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should show required asterisk when required prop is true', () => {
    const wrapper = mount(AppSelect, {
      props: {
        modelValue: null,
        label: 'Test Select',
        options,
        required: true,
      },
    })

    expect(wrapper.find('.text-red-5').exists()).toBe(true)
  })

  it('should emit update:modelValue when option is selected', async () => {
    const wrapper = mount(AppSelect, {
      props: {
        modelValue: null,
        label: 'Test Select',
        options,
      },
    })

    await wrapper.findComponent({ name: 'q-select' }).vm.$emit('update:modelValue', options[0])

    expect(wrapper.emitted('update:modelValue')?.[0]).toEqual([options[0]])
  })

  it('should filter options based on input', async () => {
    const wrapper = mount(AppSelect, {
      props: {
        modelValue: null,
        label: 'Test Select',
        options,
      },
    })

    const qSelect = wrapper.findComponent({ name: 'q-select' })
    await qSelect.vm.$emit('filter', 'Opção 1', (cb: any) => cb())
    await flushPromises()

    const vm = wrapper.vm as any
    expect(vm.filteredOptions).toHaveLength(1)
    expect(vm.filteredOptions[0].label).toBe('Opção 1')
  })
})
