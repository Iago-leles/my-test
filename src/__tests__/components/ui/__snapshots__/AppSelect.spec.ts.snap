// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppSelect > should match snapshot 1`] = `
"<div><span class="text-subtitle1 text-grey-6 text-weight-bold">Test Select <!--v-if--></span><label class="q-field row no-wrap items-start q-field--standard q-select q-field--auto-height q-select--with-input q-select--without-chips q-select--single" style="width: 100%;" for="f_5c8c637b-dca6-44fb-9d65-9f8102f08e6d">
    <!---->
    <div class="q-field__inner relative-position col self-stretch">
      <div class="q-field__control relative-position row no-wrap" tabindex="-1">
        <div class="q-field__control-container col relative-position row no-wrap q-anchor--skip">
          <div class="q-field__native row items-center"><input class="q-field__input q-placeholder col" type="search" tabindex="0" role="combobox" aria-readonly="false" aria-autocomplete="list" aria-expanded="false" aria-controls="f_5c8c637b-dca6-44fb-9d65-9f8102f08e6d_lb" id="f_5c8c637b-dca6-44fb-9d65-9f8102f08e6d" value=""></div>
          <!---->
        </div>
        <div class="q-field__append q-field__marginal row no-wrap items-center q-anchor--skip"><i class="q-icon notranslate material-icons q-select__dropdown-icon" aria-hidden="true" role="presentation">arrow_drop_down</i></div>
        <!---->
      </div>
      <!---->
    </div>
    <!---->
  </label></div>"
`;
